src/main/java/org/opensearch/knn/index/engine/KNNMethodContext.java

```diff
@@ -34,17 +35,45 @@
  * KNNMethodContext will contain the information necessary to produce a library index from an Opensearch mapping.
  * It will encompass all parameters necessary to build the index.
  */
-@AllArgsConstructor
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
 @Getter
 public class KNNMethodContext implements ToXContentFragment, Writeable {
 
     @NonNull
-    private final KNNEngine knnEngine;
+    private KNNEngine knnEngine;
     @NonNull
     @Setter
     private SpaceType spaceType;
     @NonNull
     private final MethodComponentContext methodComponentContext;
+    private boolean isEngineConfigured;
+
+    /**
+     * Copy constructor. Useful for creating a deep copy of a {@link KNNMethodContext}. Note that the engine and
+     * space type should be set.
+     *
+     * @param knnMethodContext original {@link KNNMethodContext}. Must NOT be null
+     */
+    public KNNMethodContext(KNNMethodContext knnMethodContext) {
+        if (knnMethodContext == null) {
+            throw new IllegalArgumentException("KNNMethodContext cannot be null");
+        }
+
+        this.knnEngine = knnMethodContext.knnEngine;
+        this.spaceType = knnMethodContext.spaceType;
+        this.isEngineConfigured = true;
```

Comment:
It will only be not configured if it is built from parsing. Otherwise, it will be configured.

Reply:
Issue is with respect to BWC. Currently, in the parse method, we default to nmslib. So, we cannot swap this to null or else it might break BWC. For SpaceType, which is similarly defined, @heemin32 added an "SpaceType.UNDEFINED" option. I decided to not do this and just set whether it was explicitly configured on parse in order to check if we need to update.

Reply:
probably add a comment why its introduced and cannot be null